File: com/eleet/dragonconsole/util/ANSI.java
/*
 * Copyright (c) 2010 3l33t Software Developers, L.L.C.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.eleet.dragonconsole.util;

import java.awt.Color;
import java.util.ArrayList;
import javax.swing.text.SimpleAttributeSet;
import javax.swing.text.Style;
import javax.swing.text.StyleConstants;

/** 
 * ANSI is a helper class that will process and create a
 * SimpleAttributeSet that has the color settings from a given String and
 * return it to the calling function. This is used to allow the DragonConsole
 * compatability with ANSI.
 *
 * This method will also convert DragonConsole color codes into ANSI compatible
 * codes (although if no compatible color can be found it will use the default
 * colors.
 * @author Brandon E Buck
 */
public class ANSI {
    public static final String ESCAPE = "\033"; // ANSI Escape Character that starts commands

    public static final Color BLACK = Color.BLACK;           
    public static final Color RED = Color.RED.darker();    
    public static final Color GREEN = Color.GREEN.darker();  
    public static final Color YELLOW = Color.YELLOW.darker(); 
    public static final Color BLUE = new Color(66, 66, 255).darker();  
    public static final Color MAGENTA = Color.MAGENTA.darker();
    public static final Color CYAN = Color.CYAN.darker();   
    public static final Color WHITE = Color.GRAY.brighter(); 

    public static final Color INTENSE_BLACK = Color.GRAY.darker(); 
    public static final Color INTENSE_RED = Color.RED;             
    public static final Color INTENSE_GREEN = Color.GREEN;           
    public static final Color INTENSE_YELLOW = Color.YELLOW;          
    public static final Color INTENSE_BLUE = new Color(66, 66, 255);
    public static final Color INTENSE_MAGENTA = Color.MAGENTA;         
    public static final Color INTENSE_CYAN = Color.CYAN;            
    public static final Color INTENSE_WHITE = Color.WHITE;           

    private static final Color normal[] = {BLACK, RED, GREEN, YELLOW, BLUE,
            MAGENTA, CYAN, WHITE};
    private static final Color bright[] = {INTENSE_BLACK, INTENSE_RED, 
            INTENSE_GREEN, INTENSE_YELLOW, INTENSE_BLUE, INTENSE_MAGENTA,
            INTENSE_CYAN, INTENSE_WHITE};

    /** 
     * Takes an ANSI Code as a String and breaks it apart and then creates a
     * SimpleAttributeSet with the proper Foreground and Background color.
     * @param old The current ANSI Style so any changes not specified are
     *  carried over.
     * @param string The String containing the ANSI code that needs to be
     *  processed.
     * @param defaultStyle The Default Style for the Console, used for ANSI
     *  codes 39 and 49.
     * @return Returns the SimpleAttributeSet with the proper colors of the
     *  ANSI code.
     */
    public static SimpleAttributeSet getANSIAttribute(SimpleAttributeSet old,
            String string, Style defaultStyle) {
        SimpleAttributeSet ANSI = old;

        if (ANSI == null)
            ANSI = new SimpleAttributeSet();
        
        if (string.length() > 3) {
            string = string.substring(2); // Cut off the "\033[";
            string = string.substring(0, string.length() - 1); // Remove the "m" from the end

        } else
            return null;

        String codes[] = string.split(";");

        boolean brighter = false;
        for (int i = 0; i < codes.length; i++) {

            if (codes[i].matches("[\\d]*")) {
                int code = Integer.parseInt(codes[i]);

                switch (code) {
                    case 0:
                        brighter = false;
                        ANSI = new SimpleAttributeSet();
                        break;
                    case 1:
                        brighter = true;
                        break;
                    case 30:
                    case 31:
                    case 32:
                    case 33:
                    case 34:
                    case 35:
                    case 36:
                    case 37:
                        StyleConstants.setForeground(ANSI, getColorFromANSICode(code, brighter));
                        break;
                    case 39:
                        StyleConstants.setForeground(ANSI, StyleConstants.getForeground(defaultStyle));
                        break;
                    case 40:
                    case 41:
                    case 42:
                    case 43:
                    case 44:
                    case 45:
                    case 46:
                    case 47:
                        StyleConstants.setBackground(ANSI, getColorFromANSICode(code, false));
                        break;
                    case 49:
                        StyleConstants.setBackground(ANSI, StyleConstants.getBackground(defaultStyle));
                        break;
                }
            }
        }

        return ANSI;
    }

    /** 
     * Takes a String containing a DCCC and convert it into its equivalent
     * ANSI Color Code. It does this by comparing the color associated with
     * the characters in the DCCC to the ANSI colors until it finds a match and
     * then creates the proper code, adding it to the return String. If no match
     * is made it will use the default color code 39 (for FOREGROUND) or 49
     * (for BACKGROUND).
     * @param DCCode The String containing the three character DCCC.
     * @param colors The ArrayList of TextColors that have been added to the
     *  console.
     * @return The ANSI equivalent to the given DCCC.
     */
    public static String getANSICodeFromDCCode(String DCCode, ArrayList<TextColor> colors) {
        if (DCCode.length() == 3) {
            DCCode = DCCode.substring(1); // Remove the & from the code
            char foreground = DCCode.charAt(0);
            char background = DCCode.charAt(1);
            
            String code = ESCAPE + "[0;";

            if (foreground == '0')
                code += "0";
            else if (foreground == '-')
                code += "";
            else {
                for (int i = 0; i < colors.size(); i++) {
                    TextColor tc = null;
                    if (colors.get(i).equals(TextColor.getTestTextColor(foreground))) {
                        tc = colors.get(i);
                        for (int x = 0; x < normal.length; x++) {
                            if (tc.getColor().equals(normal[x])) {
                                code += "3" + x;
                                break; // Found, no need to continue

                            } else if (tc.getColor().equals(bright[x])) {
                                code += "1;3" + x;
                                break; // Found, no need to continue
                            }
                        }
                        break; // Found, no need to continue
                    }
                }
            }

            if (code.charAt(code.length() - 1) != ';')
                code += ";";

            if (background == '0')
                code += "0";
            else if (background == '-')
                code += "";
            else {
                for (int i = 0; i < colors.size(); i++) {
                    TextColor tc = null;
                    if (colors.get(i).equals(TextColor.getTestTextColor(background))) {
                        tc = colors.get(i);
                        for (int x = 0; x < normal.length; x++) {
                            if (tc.getColor() == normal[x]) {
                                code += "4" + x;
                                break; // Found, no need to continue

                            } else if (tc.getColor().equals(bright[x])) {
                                code += "4" + x;
                                break; // Found, no need to continue
                            }
                        }
                        break; // Found, no need to continue
                    }
                }
            }

            if (code.charAt(code.length() - 1) == ';')
                code = code.substring(0, code.length() - 1);

            if (code.equals(ESCAPE + "[0;"))
                return ESCAPE + "[39;49m";
            else
                return code + "m";
        }

        return ESCAPE + "[39;49m";
    }

    /** 
     * Takes a String containing multiple DCCCs (like standard output for the
     * console) and replaces the DCCC with it's equivalent ANSI Code and returns
     * the String containing the ANSI Codes. This method uses
     * <code>getANSICodeFromDCCode()</code> to convert each DCCC into it's
     * ANSI code.
     * @param string The String that the programmer wishes to convert.
     * @param colors The ArrayList of TextColors that have been added to the
     *  console.
     * @param colorCodeChar The default colorCodeChar (used for determining a
     *  DCCC).
     * @return Returns the String after each DCCC has been converted into it's
     *  ANSI equivalent.
     */
    public static String convertDCtoANSIColors(String string, ArrayList<TextColor> colors, char colorCodeChar) {
        StringBuilder buffer = new StringBuilder(string);

        for (int i = 0; i < buffer.length(); i++) {
            if (buffer.charAt(i) == colorCodeChar) {
                if ((i + 1) < buffer.length() && (buffer.charAt(i + 1) == colorCodeChar)) {
                    i += 1; // Jump past the &&

                } else {
                    String code = buffer.substring(i, (i + 3));
                    
                    code = getANSICodeFromDCCode(code, colors);
                    int length = code.length();
                    
                    buffer = buffer.replace(i, (i + 3), code);
                    i = i + length - 1;
                }
            }
        }

        return buffer.toString();
    }

    /** 
     * Takes a String containing an ANSI Code and returns it's equivalent DCCC.
     * It does this by comparing the Color associated to the ANSI code to each
     * code contained in the list of TextColors and upon a successful match it
     * pulls the Character out of the TextColor and adds it to the DCCC String
     * it returns. If no match is found it uses the character representing the
     * default style set by the console.
     * @param code The ANSI code to convert into a DCCC.
     * @param colors The ArrayList of TextColors that have been added to the
     *  console.
     * @param colorCodeChar The colorCodeChar to place at the beginning of each
     *  DCCC.
     * @param defaultStyle The Default Style set in the Console, used if no
     *  match can be found.
     * @return Returns the DCCC equivalent of the given ANSI code.
     */
    private static String getDCCodeFromANSICode(String code, ArrayList<TextColor> colors, char colorCodeChar, String defaultStyle) {
        boolean brighter = false;
        code = code.substring(2, code.length()); // Cut off the "\033[" and "m"
        char foreground = ' ';
        char background = ' ';
        
        String colorCodes[] = code.split(";");
        for (int i = 0; i < colorCodes.length; i++) {
            if (colorCodes[i].matches("[\\d]*")) {
                int colorCode = Integer.parseInt(colorCodes[i]);
                if (colorCode == 0)
                    brighter = false;
                else if (colorCode == 1)
                    brighter = true;
                else if (colorCode >= 30 && colorCode <= 39) {
                    if (colorCode == 39)
                        foreground = defaultStyle.charAt(0);
                    else if (colorCode >= 30 && colorCode <= 37) {
                        foreground = getDCCharCodeFromColor(getColorFromANSICode(colorCode, brighter), colors);
                        brighter = false;
                    }
                } else if (colorCode >= 40 && colorCode <= 49) {
                    if (colorCode == 49)
                        background = defaultStyle.charAt(1);
                    else if (colorCode >= 40 && colorCode <= 47)
                        background = getDCCharCodeFromColor(getColorFromANSICode(colorCode, false), colors);
                }
            }
        }

        if (foreground == ' ')
            foreground = defaultStyle.charAt(0);
        if (background == ' ')
            background = defaultStyle.charAt(1);
        
        return "" + colorCodeChar + foreground + background;
    }

    /** 
     * Takes a Color and finds it's corresponding TextColor in the list of
     * TextColors that have been added to the Console (if there is one) and
     * returns the <code>charCode</code> associated with the Color given. If
     * no Color can be matched then it returns a blank character (' ').
     * @param color The Color to find in the list of TextColors.
     * @param colors The ArrayList of TextColors that have been added to the
     *  console.
     * @return The <code>charCode</code> associated with the given color, or
     *  a blank char (' ') if no match is found.
     */
    private static char getDCCharCodeFromColor(Color color,
            ArrayList<TextColor> colors) {
        TextColor test = TextColor.getTestTextColor(color);

        for (int i = 0; i < colors.size(); i++) {
            if (colors.get(i).equals(test)) {
                return colors.get(i).getCharCode();
            }
        }

        return ' ';
    }

    /** 
     * Takes a String containing ANSI Codes and converts replaces the ANSI code
     * with it's equivalent DCCC.
     * @param string The String containing then ANSI code the programmer wishes
     *  to convert to DCCCs.
     * @param colors The ArrayList of TextColors that have been added to the
     *  Console.
     * @param colorCodeChar The <code>colorCodeChar</code> set in the
     *  console.
     * @param defaultStyle The two character Default Style set in the console.
     * @return Returns the String after each ANSI Code has been converted to
     *  it's DCCC equivalent.
     */
    public static String convertANSIToDCColors(String string, 
            ArrayList<TextColor> colors, char colorCodeChar,
            String defaultStyle) {
        StringBuilder buffer = new StringBuilder(string);

        for (int i = 0; i < buffer.length(); i++) {
            if (buffer.charAt(i) == '\033') {
                if (buffer.indexOf("m", i) < buffer.length()) {
                    String code = buffer.substring(i, buffer.indexOf("m", i)); // The Color Code
                    int end = i + code.length() + 1;
                    code = getDCCodeFromANSICode(code, colors, colorCodeChar,
                            defaultStyle);
                    
                    buffer = buffer.replace(i, end, code);
                    i = i + 2;
                }
            }
        }

        return buffer.toString();
    }

    /** 
     * This method returns the Color associated with the given ANSI Code, it
     * also takes a boolean to test for color intensity. Returns null if
     * no color is found.
     * @param code The ANSI Code the color is requested for.
     * @param brighter <code>true</code> if intensity is set to bold or
     *  <code>false</code> if not.
     * @return Returns the Color associated with the ANSI code or null if none
     *  is found.
     */
    private static Color getColorFromANSICode(int code, boolean brighter) {
        switch (code) {
            case 30:
            case 40:
                if (brighter)
                    return INTENSE_BLACK;
                else
                    return BLACK;
            case 31:
            case 41:
                if (brighter)
                    return INTENSE_RED;
                else
                    return RED;
            case 32:
            case 42:
                if (brighter)
                    return INTENSE_GREEN;
                else
                    return GREEN;
            case 33:
            case 43:
                if (brighter)
                    return INTENSE_YELLOW;
                else
                    return YELLOW;
            case 34:
            case 44:
                if (brighter)
                    return INTENSE_BLUE;
                else
                    return BLUE;
            case 35:
            case 45:
                if (brighter)
                    return INTENSE_MAGENTA;
                else
                    return MAGENTA;
            case 36:
            case 46:
                if (brighter)
                    return INTENSE_CYAN;
                else
                    return CYAN;
            case 37:
            case 47:
                if (brighter)
                    return INTENSE_WHITE;
                else
                    return WHITE;
            default:
                return null;
        }
    }
}

File: src/main/kotlin/dev/bbuck/dragonconsole/text/Ansi.kt
package dev.bbuck.dragonconsole.text

import java.awt.Color
import javax.swing.text.SimpleAttributeSet
import javax.swing.text.Style
import javax.swing.text.StyleConstants

/** Provides utilities for converting to and from ANSI color codes and DragonConsole color codes. */
object ANSI {
    /** ANSI escape code that begins ANSI control codes. */
    const val ESCAPE = '\u001b'

    // Bright ANSI standard colors
    @JvmField public val INTENSE_BLACK = Color.GRAY.darker()
    @JvmField public val INTENSE_RED = Color.RED
    @JvmField public val INTENSE_GREEN = Color.GREEN
    @JvmField public val INTENSE_YELLOW = Color.YELLOW
    @JvmField public val INTENSE_BLUE = Color(66, 66, 255)
    @JvmField public val INTENSE_MAGENTA = Color.MAGENTA
    @JvmField public val INTENSE_CYAN = Color.CYAN
    @JvmField public val INTENSE_WHITE = Color.WHITE

    // Colors associated to ANSI standard colors
    @JvmField public val BLACK = Color.BLACK
    @JvmField public val RED = INTENSE_RED.darker()
    @JvmField public val GREEN = INTENSE_GREEN.darker()
    @JvmField public val YELLOW = INTENSE_YELLOW.darker()
    @JvmField public val BLUE = INTENSE_BLUE.darker()
    @JvmField public val MAGENTA = INTENSE_MAGENTA.darker()
    @JvmField public val CYAN = INTENSE_CYAN.darker()
    @JvmField public val WHITE = Color.GRAY.brighter()

    /** Normal colors in ANSI color order (0 -> Black, 1 -> Red, etc...) */
    @JvmField
    public val NORMAL_COLORS = listOf(BLACK, RED, GREEN, YELLOW, BLUE, MAGENTA, CYAN, WHITE)

    /** Intense colors in ANSI color order (0 -> Black, 1 -> Red, etc...) */
    @JvmField
    public val INTENSE_COLORS =
            listOf(
                    INTENSE_BLACK,
                    INTENSE_RED,
                    INTENSE_GREEN,
                    INTENSE_YELLOW,
                    INTENSE_BLUE,
                    INTENSE_MAGENTA,
                    INTENSE_CYAN,
                    INTENSE_WHITE
            )

    /** Signifies that following colors should be "bright" or "intense" */
    const val BRIGHT_CODE = 1

    /** Resets the current ANSI color settings back to the default. */
    const val RESET = 0

    /** Special color that represents the "default" color. */
    const val DEFAULT_COLOR_CODE = 9

    /** Color codes are always in 0 to 7. */
    val VALID_COLOR_RANGE = 0..7

    /** Foreground codes are 30-37, but we only care about the digit. */
    const val FOREGROUND_DIGIT = 3

    /** Background codes are 40-47, but we only care about the digit. */
    const val BACKGROUND_DIGIT = 4

    /** Numeric regular expression for filtering out ANSI codes not needed. */
    val NUMERIC_CODE = """\A\d+\z""".toRegex()

    /**
     * ANSI code to use the default foreground and background, used as a default for empty/incorrect
     * DCCode provided for conversion.
     */
    const val USE_DEFAULTS_CODE = "$ESCAPE[39;49m"

    /**
     * Convert the string ANSI code into a `SimpleAttributeSet` for use in a `StyledDocument`. This
     * method only considers the ANSI color code portions and ignores the rest of the code. This
     * method does not validate the code.
     *
     * @param existingAttributeSet is the current active document style as ANSI codes modify
     * existing styles instead of replacing everything about them.
     * @param ansiCode is the actual string representation of current ANSI code.
     * @param defaultStyle is the base style of the document we pull from when the ANSI code refers
     * to reverting to the default values.
     * @return the modified `SimpleAttributeSet` with the new current ANSI color configuration.
     *
     * @since 3.0
     * @author Brandon Buck
     */
    @JvmStatic
    public fun getANSIAttribute(
            existingAttributeSet: SimpleAttributeSet?,
            ansiCode: String,
            defaultStyle: Style
    ): SimpleAttributeSet? {
        if (ansiCode.length <= 3) {
            return null
        }

        var ansiAttributeSet = existingAttributeSet

        if (ansiAttributeSet == null) {
            ansiAttributeSet = SimpleAttributeSet()
        }

        var codeParts = listOf<Int>()
        // minimum code length would be the escape, brace, and 'm' (end of code)
        if (ansiCode.length > 3) {
            // trims the escape and opne character as well as the final 'm' ending
            // the code
            codeParts =
                    ansiCode.substring(2, ansiCode.length - 1)
                            .split(';')
                            .filter { it.matches(NUMERIC_CODE) }
                            .map(String::toInt)
        }

        var brightColors = false
        for (codePart in codeParts) {
            if (codePart == 0) {
                brightColors = false
                ansiAttributeSet = SimpleAttributeSet()
            } else if (codePart == 1) {
                brightColors = true
            }

            val tensDigit = codePart / 10
            val onesDigit = codePart % 10

            if (onesDigit in VALID_COLOR_RANGE) {
                val color = getColorFromAnsiCode(codePart, brightColors)
                when (tensDigit) {
                    FOREGROUND_DIGIT -> StyleConstants.setForeground(ansiAttributeSet, color)
                    BACKGROUND_DIGIT -> StyleConstants.setBackground(ansiAttributeSet, color)
                }
            } else if (onesDigit == DEFAULT_COLOR_CODE) {
                if (tensDigit == FOREGROUND_DIGIT) {
                    val color = StyleConstants.getForeground(defaultStyle)
                    StyleConstants.setBackground(ansiAttributeSet, color)
                } else if (tensDigit == BACKGROUND_DIGIT) {
                    val color = StyleConstants.getBackground(defaultStyle)
                    StyleConstants.setBackground(ansiAttributeSet, color)
                }
            }
        }

        return ansiAttributeSet
    }

    /**
     * Converts a given DCCode into an ANSI color code based on the color values of the current
     * theme. If no match is found or an invalid code is provided then a default "use default color"
     * (`[39;49m`) is returned.
     *
     * @param dcCode is the code that should be converted to an ANSI code.
     * @param colors is the list of colors currently present in the DragonConsole theme.
     * @return the best guess at an ANSI color code equivalent to the provided DCCode.
     *
     * @since 3.0
     * @author Brandon Buck
     */
    @JvmStatic
    public fun getANSICodeFromDCCode(dcCode: String, colors: List<TextColor>): String {
        if (dcCode.length != 3) {
            return USE_DEFAULTS_CODE
        }

        val colorMap = colors.associate { it.charCode to it }

        // DCCode format is &fb and we only care about the foreground and background
        // characters
        val foreground = dcCode[1]
        val background = dcCode[2]

        val ansiCode = mutableListOf("0")

        if (colorMap.containsKey(foreground)) {
            val textColor = colorMap.getValue(foreground)
            var bright = false
            var colorIndex = NORMAL_COLORS.indexOf(textColor.color)
            if (colorIndex < 0) {
                bright = true
                colorIndex = INTENSE_COLORS.indexOf(textColor.color)
            }

            if (colorIndex >= 0) {
                if (bright) {
                    ansiCode.add("1")
                }
                ansiCode.add("$FOREGROUND_DIGIT$colorIndex")
            }
        }

        if (colorMap.containsKey(background)) {
            val textColor = colorMap.getValue(background)
            var bright = false
            var colorIndex = NORMAL_COLORS.indexOf(textColor.color)
            if (colorIndex < 0) {
                bright = true
                colorIndex = INTENSE_COLORS.indexOf(textColor.color)
            }

            if (colorIndex >= 0) {
                if (bright) {
                    ansiCode.add("1")
                }
                ansiCode.add("$BACKGROUND_DIGIT$colorIndex")
            }
        }

        val generatedCode =
                ansiCode.joinToString(separator = ";", prefix = "$ESCAPE[", postfix = "m")
        if (generatedCode == "[0m") {
            return USE_DEFAULTS_CODE
        } else {
            return generatedCode
        }
    }

    /**
     * Convert all the DCCode characters in a block of text into corresponding ANSI color codes and
     * return the new string with ANSI color codes in it.
     *
     * @param text is the text contianing DCCodes.
     * @param colors are all the colors currently defined in the DragonConsole.
     * @param colorCodeChar is the character that denotes a DCCode follows, typically this will be
     * `&`.
     * @return the block of text provided withall DCCode's converted into rough equivalents of ANSI
     * codes.
     *
     * @since 3.0
     * @author Brandon Buck
     */
    @JvmStatic
    public fun convertDCtoANSIColors(
            text: String,
            colors: List<TextColor>,
            colorCodeChar: Char
    ): String {
        val output = StringBuilder()

        var i = 0
        while (i < text.length) {
            if (text[i] == colorCodeChar) {
                if (i < text.length - 1 && text[i + 1] == colorCodeChar) {
                    // double color code characters print a single color code
                    // character
                    output.append(colorCodeChar).append(colorCodeChar)
                    i += 2
                    continue
                }

                if (i < text.length - 2) {
                    val code = text.substring(i, i + 3)

                    val ansiCode = getANSICodeFromDCCode(code, colors)
                    output.append(ansiCode)
                    i += 3
                    continue
                }
            }

            output.append(text[i])
            ++i
        }

        return output.toString()
    }

    /**
     * Convert the ANSI color codes in the given text to DragonConsole color codes.
     *
     * @param text contains the ANSI color codes that need to be converted.
     * @param colors are the set of colors currently added to the DragonConsole.
     * @param colorCodeChar is the character that begins a DragonConsole color code.
     * @param defaultStyle is the default foreground and background color.
     * @return the text with all ANSI color codes converted into DragonConsole color codes.
     *
     * @since 3.0
     * @author Brandon Buck
     */
    @JvmStatic
    public fun convertANSIToDCColors(
            text: String,
            colors: List<TextColor>,
            colorCodeChar: Char,
            defaultStyle: String
    ): String {
        val output = StringBuilder()

        var i = 0
        while (i < text.length) {
            if (text[i] == ESCAPE) {
                val endIndex = text.indexOf('m', i + 1)
                if (endIndex >= 0) {
                    val ansiCode = text.substring(i, endIndex + 1)
                    val dcCode =
                            getDCCodeFromAnsiCode(ansiCode, colors, colorCodeChar, defaultStyle)
                    output.append(dcCode)
                    i = endIndex + 1
                    continue
                }
            }

            output.append(text[i])
            ++i
        }

        return output.toString()
    }

    /**
     * Converts a given ANSI code into it's DCCode equivalent based on the current defined colors in
     * the console.
     *
     * @param code is the ANSI code that needs to be converted.
     * @param colors are all the colors defined in the Console.
     * @param colorCodeChar is the character that denotes the beginning fo a DCCode, which is
     * typically `&`.
     * @param defaultStyle is the "default" DCCode to pull color characters from as needed.
     * @return the DCCode equivalent (based on color value) of the input ANSI code.
     *
     * @since 3.0
     * @author Brandon Buck
     */
    private fun getDCCodeFromAnsiCode(
            ansiCode: String,
            colors: List<TextColor>,
            colorCodeChar: Char,
            defaultStyle: String
    ): String {
        var bright = false
        val ansiCodeContents = ansiCode.substring(2)
        var foreground = defaultStyle[0]
        var background = defaultStyle[1]

        val ansiParts = ansiCodeContents.split(";")
        for (part in ansiParts) {
            if (part.matches("""\d+""".toRegex())) {
                var colorCode = part.toInt()
                when (colorCode) {
                    0 -> {
                        bright = false
                    }
                    1 -> {
                        bright = true
                    }
                    in 30..37 -> {
                        foreground =
                                getDCCharCodeFromColor(
                                        getColorFromAnsiCode(colorCode, bright),
                                        colors
                                )
                    }
                    39 -> {
                        foreground = defaultStyle[0]
                    }
                    in 40..47 -> {
                        background =
                                getDCCharCodeFromColor(
                                        getColorFromAnsiCode(colorCode, bright),
                                        colors
                                )
                    }
                    49 -> {
                        background = defaultStyle[1]
                    }
                }
            }
        }

        return "$colorCodeChar$foreground$background"
    }

    /**
     * Find the corresponding char code associated with a given color value based on what colors are
     * defined in the console.
     *
     * @param color is the search color, the color for which a character is desired.
     * @param colors are all the colors currently defined in the console.
     * @return the character representing the input color or simply an empty character.
     *
     * @since 3.0
     * @author Brandon Buck
     */
    private fun getDCCharCodeFromColor(color: Color, colors: List<TextColor>): Char {
        for (textColor in colors) {
            if (textColor.color == color) {
                return textColor.charCode
            }
        }

        return ' '
    }

    /**
     * Fetch the actual color specified by the given ANSI code.
     *
     * Valid colors fall in the range of 0 to 7 (with a 3 or 4 in the tens digit). Each of these
     * values directly corresponds to a color in ANSI so we fetch and return the correct color for
     * the code.
     *
     * @param code is the numeric ANSI code representing the color.
     * @param bright determines if the normal or intense color list should be used.
     * @return the associated color.
     *
     * @since 3.0
     * @author Brandon Buck
     */
    private fun getColorFromAnsiCode(code: Int, bright: Boolean): Color {
        val colorCode = code % 10

        if (colorCode !in VALID_COLOR_RANGE) {
            throw IllegalArgumentException("Invalid ANSI color code '$code' provided")
        }

        if (bright) {
            return INTENSE_COLORS[colorCode]
        }

        return NORMAL_COLORS[colorCode]
    }
}
